libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -3274,6 +3282,44 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::bytesToBytesNNConversionFunction(ArrayType const& _from, FixedBytesType const& _to)
+{
+	solAssert(_from.isByteArray(), "");
+	solAssert(_from.isDynamicallySized(), "");
+	string functionName = "convert_bytes_to_bytesNN_from_" + _from.identifier() + "_to_" + _to.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(array<?fromCalldata>, len</fromCalldata>) -> value {
+				let length := <arrayLen>(array<?fromCalldata>, len</fromCalldata>)
+				if gt(length, <fixedBytesLen>) { <panic>() }
+				let dataArea := array
+				<?fromMemory>
+					dataArea := <dataLocation>(array)
+				</fromMemory>
+				<?fromStorage>
+					if eq(length, 32) { dataArea := <dataLocation>(array) }
+				</fromStorage>
+				value := <extractValue>(dataArea)
+			}
+		)");
+		templ("functionName", functionName);
+		templ("fromCalldata", _from.dataStoredIn(DataLocation::CallData));
+		templ("arrayLen", arrayLengthFunction(_from));
+		templ("fixedBytesLen", to_string(_to.numBytes()));
+		templ("panic", panicFunction(PanicCode::ResourceError));
+		templ("fromMemory", _from.dataStoredIn(DataLocation::Memory));
+		templ("fromStorage", _from.dataStoredIn(DataLocation::Storage));
+		templ("dataLocation", arrayDataAreaFunction(_from));
+		templ(
+			"extractValue",
+			_from.dataStoredIn(DataLocation::Storage) ?
+			readFromStorage(_to, 32 - _to.numBytes(), true) :
```

Comment:
I think if the value is shorter than 32 bytes but we convert it to bytes32, then the last byte is not cleared.

Reply:
I think this stands for memory as well. If we have bytes array of length 16 and convert to bytes32, it will read dirty bytes from memory...
I am not sure I can use `readFrom...` helpers at all here, because bytes that are going over the length of array should be cleared.